endpoints/v2/manifest.py

```diff
@@ -47,129 +49,158 @@
 MANIFEST_TAGNAME_ROUTE = BASE_MANIFEST_ROUTE.format(VALID_TAG_PATTERN)
 
 
-@v2_bp.route(MANIFEST_TAGNAME_ROUTE, methods=["GET"])
-@disallow_for_account_recovery_mode
-@parse_repository_name()
-@process_registry_jwt_auth(scopes=["pull"])
-@require_repo_read
-@anon_protect
-def fetch_manifest_by_tagname(namespace_name, repo_name, manifest_ref):
-    repository_ref = registry_model.lookup_repository(namespace_name, repo_name)
-    if repository_ref is None:
-        image_pulls.labels("v2", "tag", 404).inc()
-        raise NameUnknown()
+class FetchManifest(MethodView):
+    decorators = [
+        anon_protect,
+        require_repo_read,
+        process_registry_jwt_auth(scopes=["pull"]),
+        parse_repository_name(),
+        disallow_for_account_recovery_mode,
+    ]
 
-    tag = registry_model.get_repo_tag(repository_ref, manifest_ref)
-    if tag is None:
-        if registry_model.has_expired_tag(repository_ref, manifest_ref):
-            logger.debug(
-                "Found expired tag %s for repository %s/%s", manifest_ref, namespace_name, repo_name
-            )
-            msg = (
-                "Tag %s was deleted or has expired. To pull, revive via time machine" % manifest_ref
-            )
+    def get(self, namespace_name, repo_name, manifest_ref):
+        try:
+            proxy = setup_proxy(namespace_name, repo_name)
+            media_type = request.headers.get("Accept", None)
+            resp = proxy.get_manifest(manifest_ref, media_type)
+            return Response(**resp)
+        except ProxyNotSupported as e:
+            logger.debug(f"Skipping pull through proxy cache: {e}")
+
+        return self.get_manifest_by_ref(namespace_name, repo_name, manifest_ref)
+
+    def get_manifest_by_ref(namespace_name, repo_name, manifest_ref):
+        raise NotImplementedError()
+
+
+class FetchManifestByTag(FetchManifest):
+    def get_manifest_by_ref(self, namespace_name, repo_name, manifest_ref):
```

Comment:
Yea... The semantics of this class and the method is weird and confusing IMO. I **really** think inheritance isn't the right abstraction in this case (I'm not a fan of inheritance in general, but that's another story).

For sharing proxy logic between manifest endpoints, see my other comment about using a decorator.

Reply:
This is pretty standard python code. I wouldn't aim at avoiding inheritance while coding in an object-oriented programming language, there's nothing wrong with it if we use it correctly.

Using inheritance here is, as you point out, a way of abstracting away the proxy code. And I've been arguing against us doing that prematurely - so I rolled these views back into functions, and added the proxy code within them, similar to the blob endpoint functions. Now we don't have any view related abstractions, so we're hopefully free of any bias towards classes or decorators.